pallets/capacity/src/lib.rs

```diff
@@ -1076,39 +1080,41 @@ impl<T: Config> Pallet<T> {
 		>,
 		DispatchError,
 	> {
-		if !Self::has_unclaimed_rewards(account) {
+		let current_era_info = Self::get_current_era();
+		let staking_history =
+			Self::get_staking_history_for(account).ok_or(Error::<T>::NotAProviderBoostAccount)?;
+
+		if !Self::has_unclaimed_rewards_with_params(current_era_info.era_index, &staking_history) {
 			return Ok(BoundedVec::new());
 		}
 
-		let staking_history =
-			Self::get_staking_history_for(account).ok_or(Error::<T>::NotAProviderBoostAccount)?; // cached read
-
-		let current_era_info = Self::get_current_era(); // cached read, ditto
 		let max_history: u32 = T::ProviderBoostHistoryLimit::get();
 
-		let mut reward_era = current_era_info.era_index.saturating_sub((max_history).into());
+		let start_era = current_era_info.era_index.saturating_sub((max_history).into());
 		let end_era = current_era_info.era_index.saturating_sub(One::one());
 
 		// start with how much was staked in the era before the earliest for which there are eligible rewards.
 		let mut previous_amount: BalanceOf<T> =
-			staking_history.get_amount_staked_for_era(&(reward_era.saturating_sub(1u32.into())));
+			staking_history.get_amount_staked_for_era(&(start_era.saturating_sub(1u32.into())));
 
 		let mut unclaimed_rewards: BoundedVec<
 			UnclaimedRewardInfo<BalanceOf<T>, BlockNumberFor<T>>,
 			T::ProviderBoostHistoryLimit,
 		> = BoundedVec::new();
-		while reward_era.le(&end_era) {
+
+		let mut chunk_cache = BTreeMap::new();
+
```

Comment:
Any advantage of using of Btree of hashmap for temporary cache ?

Reply:
Good question! I had to double-check this myself. The main reason for using BTreeMap instead of HashMap is that HashMap can’t be used in a no_std environment. [Here’s a detailed explanation](https://stackoverflow.com/questions/65469712/what-do-i-replace-vec-and-hashset-with-in-a-no-std-environmement#:~:text=You%20can't%20use%20HashMap,this%20security%20invariant%20is%20important) on why this is the case.

Reply:
Aah yes the wasm